gateway/modules/crud/batcher.go

```diff
@@ -0,0 +1,129 @@
+				if batch[project] == nil {
+					batch[project] = map[string]map[string]batchChannels{dbAlias: {tableName: {request: addInsertToBatchCh, closeC: closeC}}}
+					continue
+				}
+				batch[project][dbAlias][tableName] = batchChannels{request: addInsertToBatchCh, closeC: closeC}
+			}
+		}
+	}
+	m.batchMapTableToChan = batch
+}
+
+func (m *Module) insertBatchExecutor(done chan struct{}, addInsertToBatchCh batchRequestChan, batchTime int, project, dbAlias, tableName string, batchRecordLimit int) {
+	responseChannels := make([]batchResponseChan, 0)
+	batchRequests := make([]interface{}, 0)
+	if batchTime <= 0 { // when new project is created set default time to 100 milli seconds
+		batchTime = 100
+	}
+	if batchRecordLimit <= 0 {
+		batchRecordLimit = 100 // when new project is created set default batch record limit to 100
+	}
+	ticker := time.NewTicker(time.Duration(batchTime) * time.Millisecond)
+	for {
+		select {
+		case <-done:
+			ticker.Stop()
+			// safe operation since SetConfig will hold a lock preventing others from writing into this channel after its closed
+			close(addInsertToBatchCh)
+			close(done)
+			logrus.Debugf("closing batcher for database %s table %s", dbAlias, tableName)
+			return
+		case v := <-addInsertToBatchCh:
+			responseChannels = append(responseChannels, v.response)
+			switch docType := v.document.(type) {
+			case map[string]interface{}:
+				batchRequests = append(batchRequests, docType)
+			case []interface{}:
+				batchRequests = append(batchRequests, docType...)
+			}
+			if len(batchRequests) >= batchRecordLimit {
+				m.executeBatch(project, dbAlias, tableName, batchRequests, responseChannels)
+				batchRequests = make([]interface{}, 0)          // clear the requests array
+				responseChannels = make([]batchResponseChan, 0) // clear the response channels array
+				// reset ticker
+				ticker.Stop()
+				ticker = time.NewTicker(time.Duration(batchTime) * time.Millisecond)
+			}
+		case <-ticker.C:
+			if len(batchRequests) > 0 {
+				m.executeBatch(project, dbAlias, tableName, batchRequests, responseChannels)
+				batchRequests = make([]interface{}, 0)          // clear the requests array
+				responseChannels = make([]batchResponseChan, 0) // clear the response channels array
+			}
+		}
+	}
+}
+
+func (m *Module) executeBatch(project, dbAlias, tableName string, batchRequests []interface{}, responseChannels []batchResponseChan) {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	if err := m.InternalCreate(ctx, dbAlias, project, tableName, &model.CreateRequest{Operation: utils.All, Document: batchRequests}); err != nil {
+		logrus.Errorf("error executing batch request for database %s table %s - %s", dbAlias, tableName, err)
+		m.sendResponses(responseChannels, batchResponse{err: err, docsInserted: 0})
+		return
+	}
+	// send response to all client request
+	m.sendResponses(responseChannels, batchResponse{err: nil, docsInserted: int64(len(batchRequests))})
```

Comment:
The docs inserted you've put here is incorrect since this is the batch size not the number of documents inserted by individual requests

Reply:
You should simply send the err which is nil

Reply:
@YourTechBud but for metrics module we need count of docs inserted
and the batchRequests variable represents the number of recordes inserted 
why can't it represent the count ?

Reply:
ok i understood from the review given below